In [2]:
import pandas as pd
import numpy as np
# Load dataset
df = pd.read_csv("C:/Users/SMRC/Downloads/DigitalServiceUsage_Rwanda.csv")

In [3]:
# lets understand the nature of the data
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   District                1025 non-null   object 
 1   Service_Name            1025 non-null   object 
 2   Department              1025 non-null   object 
 3   Users_Reported          501 non-null    float64
 4   Satisfaction_Score_(%)  510 non-null    float64
 5   Year                    1025 non-null   int64  
 6   Month                   937 non-null    object 
dtypes: float64(2), int64(1), object(4)
memory usage: 56.2+ KB
None


In [4]:
df.head()

District              Service_Name                Department  \
0    Musanze               eTax Portal        Ministry of Health   
1     Gasabo  Land Registration Portal    Immigration Department   
2     Rusizi               eTax Portal  Rwanda Revenue Authority   
3  Nyagatare            E-Visa Service           Ministry of ICT   
4  Nyagatare    Digital Health Records  Rwanda Revenue Authority   

   Users_Reported  Satisfaction_Score_(%)  Year    Month  
0           910.0                     NaN  2022  Unknown  
1          3822.0                    93.6  2023      Dec  
2           516.0                    82.9  2024      Mar  
3          4476.0                     NaN  2024      Jun  
4             NaN                     NaN  2023      May

In [5]:
df.tail()

District              Service_Name                Department  \
1020     Kayonza  Land Registration Portal  Rwanda Revenue Authority   
1021   Nyagatare            E-Visa Service           Ministry of ICT   
1022   Rwamagana           Irembo Services  Rwanda Revenue Authority   
1023  Nyarugenge           Irembo Services        Ministry of Health   
1024    Kicukiro               eTax Portal        Ministry of Health   

      Users_Reported  Satisfaction_Score_(%)  Year Month  
1020          1115.0                     NaN  2022   Sep  
1021          1009.0                    84.8  2023   Jan  
1022             NaN                    78.9  2024   May  
1023           370.0                     NaN  2023   Oct  
1024             NaN                    77.0  2021   Nov

In [6]:
df.describe()

Users_Reported  Satisfaction_Score_(%)         Year
count      501.000000              510.000000  1025.000000
mean      2524.217565               75.161569  2022.437073
std       1431.196236               14.497803     1.119863
min         51.000000               50.100000  2021.000000
25%       1268.000000               62.300000  2021.000000
50%       2447.000000               75.450000  2022.000000
75%       3798.000000               87.700000  2023.000000
max       4983.000000              100.000000  2024.000000

In [7]:
# Checking for missing values
print(df.isnull().sum())

District                    0
Service_Name                0
Department                  0
Users_Reported            524
Satisfaction_Score_(%)    515
Year                        0
Month                      88
dtype: int64


We observe that Users_Reported is missing 524 values while Satisfaction_Score_(%) attribute miss 515. Since these attributes are considered important in analysis, the analysis decided to fill the missing values with median since they are continous in nature.

In [8]:
df['Users_Reported'].fillna(df['Users_Reported'].median(), inplace=True)
df['Satisfaction_Score_(%)'].fillna(df['Satisfaction_Score_(%)'].median(), inplace=True)

In [10]:
print(df.isnull().sum())

District                   0
Service_Name               0
Department                 0
Users_Reported             0
Satisfaction_Score_(%)     0
Year                       0
Month                     88
dtype: int64


we observe that the month attribute still have missing values. it needs to be handled indipendently. we start by checking its format

In [11]:
df['Month'].unique()

array(['Unknown', 'Dec', 'Mar', 'Jun', 'May', 'Aug', nan, 'Apr', 'Sep',
       'Feb', 'Jan', 'Nov', 'Jul', 'Oct'], dtype=object)

lets hanlde inconsistancies and then Fill by Related Group

In [12]:
df['Month'] = df['Month'].str.strip().str.title()

In [13]:
df['Month'] = df.groupby('District')['Month'].transform(
    lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 'Unknown')
)

In [14]:
print(df.isnull().sum())

District                  0
Service_Name              0
Department                0
Users_Reported            0
Satisfaction_Score_(%)    0
Year                      0
Month                     0
dtype: int64


The above output shows that all the missing values have been handled

In [15]:
# lets check duplicated values
print(df.duplicated().sum())

25


We find that there are 25 duplicates in this dataset

In [17]:
df[df.duplicated()]

District                  Service_Name                Department  \
1000      Rubavu      Land Registration Portal         Ministry of Lands   
1001     Kayonza               Irembo Services        Ministry of Health   
1002     Gakenke                E-Visa Service        Ministry of Health   
1003      Rusizi                   eTax Portal  Rwanda Revenue Authority   
1004      Nyanza                E-Visa Service           Ministry of ICT   
1005  Nyarugenge  Online Business Registration        Ministry of Health   
1006     Karongi                E-Visa Service        Ministry of Health   
1007  Nyarugenge        Digital Health Records    Immigration Department   
1008     Rulindo               Irembo Services        Ministry of Health   
1009    Bugesera                   eTax Portal         Ministry of Lands   
1010     Karongi               Irembo Services         Ministry of Lands   
1011    Kicukiro                E-Visa Service           Ministry of ICT   
1012   Rwamagana                E-Visa Service         Ministry of Lands   
1013    Bugesera                E-Visa Service         Ministry of Lands   
1014      Nyanza      Land Registration Portal        Ministry of Health   
1015      Rusizi        Digital Health Records         Ministry of Lands   
1016      Rusizi        Digital Health Records         Ministry of Lands   
1017     Kayonza               Irembo Services         Ministry of Lands   
1018     Musanze                E-Visa Service        Ministry of Health   
1019      Rubavu        Digital Health Records           Ministry of ICT   
1020     Kayonza      Land Registration Portal  Rwanda Revenue Authority   
1021   Nyagatare                E-Visa Service           Ministry of ICT   
1022   Rwamagana               Irembo Services  Rwanda Revenue Authority   
1023  Nyarugenge               Irembo Services        Ministry of Health   
1024    Kicukiro                   eTax Portal        Ministry of Health   

      Users_Reported  Satisfaction_Score_(%)  Year    Month  
1000          2447.0                   68.20  2021      Mar  
1001          1290.0                   99.70  2022      Jun  
1002          2447.0                   69.50  2022  Unknown  
1003          4458.0                   75.45  2024      Mar  
1004          2447.0                   75.45  2023      Jul  
1005          2447.0                   54.00  2021      Feb  
1006          4855.0                   80.40  2022      Oct  
1007          2447.0                   67.80  2021      Jul  
1008          3980.0                   75.45  2023      Mar  
1009          2447.0                   75.45  2022      Mar  
1010          2447.0                   94.20  2022      Jun  
1011           613.0                   75.45  2021      Mar  
1012          4472.0                   53.50  2021      Oct  
1013          2447.0                   75.45  2023      Mar  
1014          2447.0                   75.45  2024      Aug  
1015          2447.0                   79.00  2021  Unknown  
1016          2447.0                   75.45  2023      Dec  
1017          2447.0                   75.45  2023      Nov  
1018          4856.0                   57.40  2024      Jun  
1019          2447.0                   75.45  2021      Aug  
1020          1115.0                   75.45  2022      Sep  
1021          1009.0                   84.80  2023      Jan  
1022          2447.0                   78.90  2024      May  
1023           370.0                   75.45  2023      Oct  
1024          2447.0                   77.00  2021      Nov

In [18]:
# removing duplicates
df.drop_duplicates(inplace=True)

In [19]:
print(df.duplicated().sum())

0


We observe that the duplicated values have been removed

In [20]:
# Check unique values for categorical columns
for col in df.select_dtypes(include='object'):
    print(f"\n{col} unique values:\n", df[col].unique())

# Check data types
print(df.dtypes)


District unique values:
 ['Musanze' 'Gasabo' 'Rusizi' 'Nyagatare' 'Rubavu' 'Burera' 'Nyarugenge'
 'Nyamagabe' 'Kicukiro' 'Huye' 'Nyanza' 'Rulindo' 'Rwamagana' 'Karongi'
 'Musanze-' 'Gakenke' 'Bugesera' 'Ngoma' 'Kayonza' 'Nyamagabe-' 'Ngoma-'
 'Huyex' 'Karongi-' 'Karongix' 'Bugeserax' 'Musanzex' 'Kicukirox'
 'Nyanzax']

Service_Name unique values:
 ['eTax Portal' 'Land Registration Portal' 'E-Visa Service'
 'Digital Health Records' 'Online Business Registration' 'Irembo Services']

Department unique values:
 ['Ministry of Health' 'Immigration Department' 'Rwanda Revenue Authority'
 'Ministry of ICT' 'Ministry of Lands']

Month unique values:
 ['Unknown' 'Dec' 'Mar' 'Jun' 'May' 'Aug' 'Apr' 'Sep' 'Feb' 'Jan' 'Nov'
 'Jul' 'Oct']
District                   object
Service_Name               object
Department                 object
Users_Reported            float64
Satisfaction_Score_(%)    float64
Year                        int64
Month                      object
dtype: object


We observe that District column has some District that typed wrong and in Department column too there there is no ministry of lands exists here in rwanda instead there is Ministry of environment and the national land authority which is in charge of land registrations and then in Month column I decided to make it full so that anyone can easly understand what that means. 

In [21]:
df['District'].unique()

array(['Musanze', 'Gasabo', 'Rusizi', 'Nyagatare', 'Rubavu', 'Burera',
       'Nyarugenge', 'Nyamagabe', 'Kicukiro', 'Huye', 'Nyanza', 'Rulindo',
       'Rwamagana', 'Karongi', 'Musanze-', 'Gakenke', 'Bugesera', 'Ngoma',
       'Kayonza', 'Nyamagabe-', 'Ngoma-', 'Huyex', 'Karongi-', 'Karongix',
       'Bugeserax', 'Musanzex', 'Kicukirox', 'Nyanzax'], dtype=object)

In [22]:
df.loc[:,'District'] = df['District'].replace({
    'Musanze-': 'Musanze',
    'Nyamagabe-': 'Nyamagabe',
    'Ngoma-': 'Ngoma',
    'Huyex': 'Huye',
    'Karongi-': 'Karongi',
    'Karongix': 'Karongi',
    'Bugeserax': 'Bugesera',

    'Musanzex': 'Musanze',
    'Kicukirox': 'Kicukiro',
    'Nyanzax': 'Nyanza',
})

In [23]:
df['District'].unique()

array(['Musanze', 'Gasabo', 'Rusizi', 'Nyagatare', 'Rubavu', 'Burera',
       'Nyarugenge', 'Nyamagabe', 'Kicukiro', 'Huye', 'Nyanza', 'Rulindo',
       'Rwamagana', 'Karongi', 'Gakenke', 'Bugesera', 'Ngoma', 'Kayonza'],
      dtype=object)

In [24]:
df.loc[:,'Department'] = df['Department'].replace({
    'Ministry of Lands': 'National Land Authority',
})

In [25]:
df['Department'].unique()

array(['Ministry of Health', 'Immigration Department',
       'Rwanda Revenue Authority', 'Ministry of ICT',
       'National Land Authority'], dtype=object)

In [26]:
df.loc[:,'Month'] = df['Month'].map({
    
    'Unknown' :'Unknown',
    'Dec' :'December',
    'Mar' :'March',
    'Jun' :'June',
    'May' :'May',
    'Aug' :'August',
    'Apr' :'April',
    'Sep' :'September',
    'Feb' :'February',
    'Jan' :'January',
    'Nov' :'November',
    'Jul' :'July',
    'Oct' :'October',
})

In [28]:
df['Month'].unique()

array(['Unknown', 'December', 'March', 'June', 'May', 'August', 'April',
       'September', 'February', 'January', 'November', 'July', 'October'],
      dtype=object)

In [29]:
df.head()

District              Service_Name                Department  \
0    Musanze               eTax Portal        Ministry of Health   
1     Gasabo  Land Registration Portal    Immigration Department   
2     Rusizi               eTax Portal  Rwanda Revenue Authority   
3  Nyagatare            E-Visa Service           Ministry of ICT   
4  Nyagatare    Digital Health Records  Rwanda Revenue Authority   

   Users_Reported  Satisfaction_Score_(%)  Year     Month  
0           910.0                   75.45  2022   Unknown  
1          3822.0                   93.60  2023  December  
2           516.0                   82.90  2024     March  
3          4476.0                   75.45  2024      June  
4          2447.0                   75.45  2023       May

In [30]:
# Save to CSV in the same folder as the notebook
df.to_csv("DigitalServiceUsage_Rwanda_cleaned_Dataset.csv", index=False)